src/System.IO/src/System/IO/BinaryWriter.cs

```diff
@@ -71,6 +72,8 @@ protected virtual void Dispose(bool disposing)
         {
             if (disposing)
             {
+                ArrayPool<byte>.Shared.Return(_buffer);
```

Comment:
Our system is not designed to make finalization efficient.   Also once you have finalization, you typically start talking about IDisposable, which is a non-trivial semantic overhead (It is like buffer manipulation, in theory it not hard, but we screw it up all the time).    

As a general rule I don't like to see IDisposable on a class unless it is required for semantic reasons (not just for perf (unless perf the whole reason for the class)).   Finalization without IDisposable is not so bad, but makes the type heavier, and thus I would not do it without (as Stephen indicated) having some data showing it is beneficial.